google-cloud-bigtable/src/main/java/com/google/cloud/bigtable/data/v2/models/BulkMutations.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.cloud.bigtable.data.v2.models;
+
+import com.google.api.core.ApiFuture;
+import com.google.api.core.ApiFutureCallback;
+import com.google.api.core.ApiFutures;
+import com.google.api.core.InternalApi;
+import com.google.api.gax.rpc.UnaryCallable;
+import com.google.common.base.Preconditions;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicLong;
+import org.threeten.bp.Duration;
+
+/**
+ * Tracker for outstanding bulk mutations. Allows for the caller to wait for all of the outstanding
+ * mutations to complete.
+ *
+ * @see com.google.cloud.bigtable.data.v2.BigtableDataClient#newBulkMutations() for example usage.
+ */
+public final class BulkMutations implements AutoCloseable {
+  private final UnaryCallable<RowMutation, Void> callable;
+  // Shared
+  private final Object lock = new Object();
+  private final AtomicLong numOutstanding = new AtomicLong();
+  // Only accessed by grpc
+  private long numFailures;
+  // Only accessed by caller
+  private boolean closed;
+
+  @InternalApi
+  public BulkMutations(UnaryCallable<RowMutation, Void> callable) {
+    this.callable = callable;
+  }
+
+  /**
+   * Waits up to 10 minutes for all of the outstanding mutations to be resolved.
+   *
+   * @see #close(Duration) for more infortmation.
+   */
+  @Override
+  public void close() throws InterruptedException, TimeoutException {
+    close(Duration.ofMinutes(10));
+  }
+
+  /**
+   * Prevents further mutations and waits for all outstanding mutations to complete.
+   *
+   * @throws BulkMutationFailure If any mutations failed.
+   * @throws InterruptedException If interrupted.
+   * @throws TimeoutException If the outstanding requests don't finish in time.
+   */
+  public void close(Duration duration) throws InterruptedException, TimeoutException {
+    closed = true;
+    long deadlineMs = System.currentTimeMillis() + duration.toMillis();
+
+    synchronized (lock) {
+      while (numOutstanding.get() > 0) {
+        long waitMs = deadlineMs - System.currentTimeMillis();
+        if (waitMs <= 0) {
+          throw new TimeoutException("Timed out waiting outstanding mutations to finish");
+        }
+        lock.wait(waitMs);
+      }
+
+      if (numFailures > 0) {
```

Comment:
Is there a race for `numFailures` between this and `onMutationComplete`?

Reply:
I don't believe so. In `close()` `numFailures` is only checked after `numOutstanding` is zero. And in `onMutationComplete`, `numOutstanding` is only updated after updating `numFailures`.
I added some comments to clarify